drivers/dma/dma_stm32.c

```diff
@@ -570,25 +593,16 @@ static int dma_stm32_init(struct device *dev)
 
 	config->config_irq(dev);
 
-	int size_stream =
-		sizeof(struct dma_stm32_stream) * data->max_streams;
-	data->streams = k_malloc(size_stream);
```

Comment:
Glad we removed this. One headache problem this cause is that default HEAP_MEM_POOL_SIZE in zephyr is 0, we have to manually set one, which is 1024 in this driver. But this is not the right way of doing this, we should ask for some bytes to be added to the HEAP_MEM_POOL_SIZE, which is a way Zephyr not supported(don't know if this is supported yet).
Anyway, since we removed k_malloc, we should remove size definition of HEAP_MEM_POOL_SIZE on soc/arm/st_stm32/common/Kconfig.defconfig.series:78. But I see the dma_mux driver is still using the k_malloc, we can define HEAP_MEM_POOL_SIZE when DMAMUX_STM32 is defined, which is defined on drivers/dma/Kconfig.stm32:30

Reply:
Here I wasn't asking for refactoring dmamux. I also support to make it in seperate PRs. What I want to say is for the new driver, we do not need HEAP_MEM_POOL_SIZE at all, but we are still setting it to 1024. My suggestion is to keep the action of setting HEAP_MEM_POOL_SIZE to dmamux driver itself.

Reply:
Where are we setting HEAP_MEM_POOL_SIZE  ?

Reply:
I'm thinking about put it in drivers/dma/Kconfig.stm32:28, which means that HEAP_MEM_POOL_SIZE will be set to 1024 when the DMAMUX driver is selected.

Reply:
OK but there is nothing in this PR (and the DMA driver) that selects HEAP_MEM_POOL_SIZE, is there ?

Reply:
You didn't get it...   Maybe I'm not explaining this clear enough... :sweat_smile:
k_malloc in zephyr needs user to set HEAP_MEM_POOL_SIZE to a non-zero value so we can k_malloc some memory. Since this driver was using k_malloc and the default value of HEAP_MEM_POOL_SIZE in Zephyr is zero, we set HEAP_MEM_POOL_SIZE to 1024 as long as `DMA` symbol is set in ST series socs, which is done on soc/arm/st_stm32/common/Kconfig.defconfig.series:78.
But our new driver gets rid of k_malloc, thus we can get rid of action of setting HEAP_MEM_POOL_SIZE.
The ideal way would be remove it, but dmamux driver is still using k_malloc, so HEAP_MEM_POOL_SIZE have to be set when dmamux is used but leaves untouched in this driver.
This PR isn't setting HEAP_MEM_POOL_SIZE, but the original one is. This PR should let HEAP_MEM_POOL_SIZE is set to 1024 only when dmamux driver is in use.

Reply:
Ah OK now I get you. I moved the mempool selection under dmamux, this should reduce memory use on systems without dmamux which is good news for @nandojve I think.

PS: I just ordered a L4R5 based board to do the DMAMUX development.